axlearn/common/causal_lm.py

```diff
@@ -32,16 +32,242 @@
+from axlearn.common.metrics import BaseLossMetrics, WeightedScalar
+from axlearn.common.module import Module, NestedTensor, Tensor, child_context, new_output_collection
 from axlearn.common.param_init import PARAM_REGEXP_WEIGHT, DefaultInitializer, WeightInitializer
-from axlearn.common.utils import flatten_items, with_sharding_constraint
+from axlearn.common.utils import (
+    Nested,
+    flatten_items,
+    validate_contains_paths,
+    with_sharding_constraint,
+)
 
 
 def layer_norm_config(eps=1e-5):
     return LayerNorm.default_config().set(eps=eps)
 
 
+class CrossEntropyLossMetrics(BaseLossMetrics):
+    """Computes cross entropy loss and related training summaries."""
+
+    @config_class
+    class Config(BaseLossMetrics.Config):
+        """Configures CrossEntropyLossMetrics.
+
+        Attributes:
+            z_loss_scale: An auxiliary z-loss scale. If not None and >0, encourages the softmax
+                normalizer to be well behaved.
+        """
+
+        z_loss_scale: Optional[float] = None
+
+    def forward(
+        self, input_batch: Nested[Tensor], *, predict_outputs: Nested[Tensor]
+    ) -> tuple[Tensor, Nested[Tensor]]:
+        """Computes cross entropy loss.
+
+        Args:
+            input_batch: A dict containing at minimum:
+                target_labels: An int Tensor of shape [...]. Negative targets do not contribute to
+                    the loss calculation.
+            predict_outputs: A dict containing at minimum:
+                logits: A float Tensor of shape [..., num_classes].
+
+        Returns:
+            A tuple (loss, metrics):
+                loss: A scalar float Tensor corresponding to cross entropy loss, including auxiliary
+                    z-loss if `cfg.z_loss_scale` is provided.
+                metrics: A dict containing:
+                    cross_entropy: Same as loss.
+                    per_token_loss: A float Tensor of same shape as `target_labels`. Ignored targets
+                        will be masked, i.e., have per-token loss of 0.
+                    live_targets: A bool Tensor of same shape as `target_labels`. False indicates
+                        ignored targets.
+                    num_targets: A scalar int Tensor corresponding to number of live targets.
+        """
+        validate_contains_paths(input_batch, paths=["target_labels"])
+        validate_contains_paths(predict_outputs, paths=["logits"])
+
+        cfg: CrossEntropyLossMetrics.Config = self.config
+
+        target_labels: Tensor = input_batch["target_labels"]
+        target_num_bytes: Optional[Tensor] = input_batch.get("target_num_bytes")
+        logits = predict_outputs["logits"]
+
+        live_targets = target_labels >= 0
+        num_targets = live_targets.sum()
+        accuracy = (
+            jnp.equal(jnp.argmax(logits, axis=-1), target_labels) * live_targets
+        ).sum() / jnp.maximum(1, num_targets)
+
+        loss, loss_dict = cross_entropy(
```

Comment:
Hi Mark, do we need a separate accuracy calculation here (above cross_entropy) if the loss_dict from cross_entropy already contains an accuracy entry? @markblee